File: backend/apps/notifications/interfaces/i_notification_service.py
"""
Root ABC for the notification service — Observer subject.

Responsibility (SRP): declare the contract for dispatching and querying notifications.
    No channel logic, no ORM access — only orchestration signatures.
Depends on: abc.ABC — nothing from the domain.
Pattern: Singleton (the concrete NotificationService is a module-level singleton)
         + Observer subject (receives events from the ticket signal).
SOLID: DIP · OCP · SRP

Sprint coverage:
    S19 → this file (contract)
    S20 → NotificationService(INotificationService) — Singleton implementation
    S26 → useNotifications hook depends on this interface on the FE side
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class INotificationService(ABC):
    """Abstract contract for notification dispatch and management."""

    @abstractmethod
    def dispatch(self, event: dict) -> None:
        """
        Observer entry point — called by the post_save(TicketEvent) signal handler.

        Args:
            event: serialized dict of a TicketEvent (not the model instance, to avoid
                   circular import between apps.tickets and apps.notifications).
                   Keys: ticket_id, tipo_evento, estado_anterior, estado_nuevo,
                         comentario, autor_id, cliente_id, asignado_id.

        Behavior:
            1. Determines recipients based on tipo_evento (SRP: this method decides who).
            2. Loads NotificationPreference for each recipient.
            3. For each active channel: NotificationFactory.build(canal).send().
        """
        ...

    @abstractmethod
    def get_user_notifications(self, user, page: int = 1) -> dict:
        """
        Return paginated notifications for the given user.

        Returns:
            {'items': list[dict], 'total': int, 'unread_count': int, 'page': int}
        """
        ...

    @abstractmethod
    def mark_as_read(self, notification_id: int, user) -> dict:
        """
        Mark a single notification as read. Enforces ownership.

        Returns: updated notification dict.
        Raises: NotificationNotFound if not found or not owned by user.
        """
        ...

    @abstractmethod
    def get_preferences(self, user) -> dict:
        """
        Return (or create with defaults) the NotificationPreference for the user.

        Returns: {'email_activo': bool, 'in_app_activo': bool, 'ws_activo': bool}
        """
        ...

    @abstractmethod
    def set_preferences(self, user, data: dict) -> dict:
        """
        Update the NotificationPreference for the user.

        Args:
            data: partial dict with any of email_activo, in_app_activo, ws_activo.

        Returns: updated preferences dict.
        """
        ...

File: frontend/src/modules/notifications/interfaces/INotificationService.ts
/**
 * Root contract for the notification service on the frontend.
 *
 * Responsibility (SRP): declare how components and hooks interact with notifications.
 *     No channel logic, no WebSocket management — only application-level operations.
 * Depends on: Notification and NotificationPreferences types defined below.
 * Pattern: DIP anchor — useNotifications hook depends on this, never on a concrete service.
 * SOLID: DIP · OCP · SRP
 *
 * Sprint coverage:
 *   S19 → this file (contract + types)
 *   S26 → NotificationClientContext + useNotifications hook implement/consume this
 */

// ── Shared types (single source of truth: ./types.ts) ────────────────────────
import type {
  Notification,
  NotificationPreferences,
  PaginatedNotifications,
} from './types'

export type {
  NotificationTipo,
  Notification,
  NotificationPreferences,
  PaginatedNotifications,
} from './types'

// ── Service contract ──────────────────────────────────────────────────────────

export interface INotificationService {
  /**
   * Load paginated notifications for the current user.
   * Called on mount and after marking notifications as read.
   */
  getUserNotifications(page?: number): Promise<PaginatedNotifications>

  /**
   * Mark a single notification as read.
   * Returns the updated notification.
   */
  markAsRead(notificationId: string): Promise<Notification>

  /** Mark every unread notification in one server-side batch. */
  markAllAsRead(): Promise<number>

  /** Load the current user's channel preferences. */
  getPreferences(): Promise<NotificationPreferences>

  /** Partially update the current user's channel preferences. */
  setPreferences(data: Partial<NotificationPreferences>): Promise<NotificationPreferences>
}
